tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -229,38 +297,65 @@ def get_asts(self) -> List[chapel.core.AstNode]:
         with self.context.track_errors() as _:
             return self.parse_file()
 
-    def rebuild_segments(self):
+    def rebuild_index(self):
         """
         Rebuild the cached line info and siblings information
 
         Note: this is a potentially expensive operation, it should only be done
         when advancing the revision
         """
         asts = self.parse_file()
-        # get ids
-        self.segments = []
+
+        self.uses_here = defaultdict(list)
+        self.use_segments.clear()
         for node, _ in chapel.each_matching(asts, chapel.core.Identifier):
             to = node.to_node()
             if to:
-                self.segments.append(ResolvedPair(NodeAndRange(node), NodeAndRange(to)))
+                self.uses_here[to.unique_id()].append(NodeAndRange(node))
+                self.use_segments.append(ResolvedPair(NodeAndRange(node), NodeAndRange(to)))
         for node, _ in chapel.each_matching(asts, chapel.core.Dot):
             to = node.to_node()
             if to:
-                self.segments.append(ResolvedPair(NodeAndRange(node), NodeAndRange(to)))
-        self.segments.sort(key=lambda s: s.ident.rng.start)
+                self.uses_here[to.unique_id()].append(NodeAndRange(node))
+                self.use_segments.append(ResolvedPair(NodeAndRange(node), NodeAndRange(to)))
+        self.use_segments.sort()
+
+        self.def_segments.clear()
+        for node, _ in chapel.each_matching(asts, chapel.core.NamedDecl):
+            self.def_segments.append(NodeAndRange(node))
+        self.def_segments.sort()
```

Comment:
This is the third `each_matching` in this function. A way to make this more performant would be to use a visitor that does all three things in 1 pass, rather than 3 passes. A lighter weight approach would be just manually rewrite a custom `each_matching` that looks for Identifier, Dot, and NamedDecl

Reply:
Made FileInfo a visitor (seemed cleanest).